### PR TITLE
Adds key mapping support for use in destination plugins.

### PR DESCRIPTION
**Pull request by bsneed**
- Adds `mapKeys` function to JSON object w/ optional value transform step.
- Adds tests for key mapping functionality.

### CODE REVIEW

File: Sources/Segment/Utilities/JSON.swift [hunk 1]
```diff
@@ -267,9 +267,73 @@ extension JSON {
...
+        let mapped = try dict.mapKeys(keys, valueTransform: valueTransform)
+        let result = try JSON(mapped)
+        return result
+    }
+}
 
 // MARK: - Helpers
 
+extension Dictionary where Key == String, Value == Any {
+    internal func mapKeys(_ keys: [String: String], valueTransform: ((_ key: Key, _ value: Value) -> Any)? = nil) throws -> [Key: Value] {
```
Comment (migs647) [inline Sources/Segment/Utilities/JSON.swift:288]: Do you need `= nil` on there?

File: Sources/Segment/Utilities/JSON.swift [hunk 2]
```diff
@@ -267,9 +267,73 @@ extension JSON {
...
+            if keys.keys.contains(key) {
+                if let mappedKey = keys[key] {
+                    // if so, lets change the key to the new value.
+                    newKey = mappedKey
+                }
+            }
+            // is this value a dictionary?
+            if let dictValue = value as? [Key: Value] {
+                if let r = try? dictValue.mapKeys(keys, valueTransform: valueTransform) {
+                    // if so, lets recurse...
```
Comment (migs647) [inline Sources/Segment/Utilities/JSON.swift:303]: 👍🏻

File: Sources/Segment/Utilities/JSON.swift [hunk 3]
```diff
@@ -267,9 +267,73 @@ extension JSON {
...
+                    newValue = r
+                }
+            } else if let arrayValue = value as? [Value] {
+                // if it's an array, we need to see if any dictionaries are within and process
```
Comment (migs647) [inline Sources/Segment/Utilities/JSON.swift:307]: 👍🏻

File: Sources/Segment/Utilities/JSON.swift [hunk 4]
```diff
@@ -267,9 +267,73 @@ extension JSON {
...
+                        }
+                    }
+                    return newValue
+                }
+            }
+            
+            if !(newValue is [Key: Value]), let transform = valueTransform {
+                // it's not a dictionary apply our transform.
+                
+                // note: if it's an array, we've processed any dictionaries inside
```
Comment (migs647) [inline Sources/Segment/Utilities/JSON.swift:323]: 👍🏻

File: Tests/Segment-Tests/JSON_Tests.swift [hunk 5]
```diff
@@ -130,4 +130,51 @@ class JSONTests: XCTestCase {
...
+            if let v = newValue as? Int {
+                if v == 1 {
+                    newValue = 11
+                }
+            }
+            print("value = \(value.self)")
+            return newValue
+        }).dictionaryValue
+        
+        XCTAssertTrue(output!["AKey1"] as! Int == 11)
```
Comment (migs647) [inline Tests/Segment-Tests/JSON_Tests.swift:167]: Can we change these to as? so we fail the unit test instead of crash the unit tests?

**Review by migs647 — APPROVED** [review]
LGTM -- just some nitpicks

**Review by alanjcharles — APPROVED** [review]
looks great. Awesome idea.

File: Sources/Segment/Utilities/JSON.swift (hunk 1 above)
Comment (bsneed) [reply Sources/Segment/Utilities/JSON.swift:288]: Yeah, cuz I want it to be optional.  They likely just want to map keys, at which point, the values won't be transformed.

File: Tests/Segment-Tests/JSON_Tests.swift (hunk 5 above)
Comment (bsneed) [reply Tests/Segment-Tests/JSON_Tests.swift:167]: We could I guess, though that doesn't really bother me.  We'd need to go hit a bunch of other tests as well as that's my MO for unit tests.  FORCE UNWRAP ALL THE THINGS!

**Pull request merged**